### PR TITLE
Fixing some performance bottlenecks

### PR DESCRIPTION
**Pull request by deedy5**
## pprofile tests

***test.py***
```python3
from glob import glob
from os.path import isdir
from charset_normalizer import detect

def performance_compare(size_coeff):
    if not isdir("./char-dataset"):
        print("This script require https://github.com/Ousret/char-dataset to be cloned on package root directory")
        exit(1)
    for tbt_path in sorted(glob("./char-dataset/**/*.*")):
        with open(tbt_path, "rb") as fp:
            content = fp.read() * size_coeff            
        detect(content)

if __name__ == "__main__":
    performance_compare(1)
```
---

### Before
```python3
pprofile --format callgrind --out cachegrind.out.original.test test.py
```
Time: 838.97 s.
[cachegrind.out.original.zip](https://github.com/Ousret/charset_normalizer/files/8598533/cachegrind.out.original.zip)
cachegrind out original test

---
### After commits
```python3
pprofile --format callgrind --out cachegrind.out.commits.test test.py
```
Time: 698.53 s.
[cachegrind.out.commits.zip](https://github.com/Ousret/charset_normalizer/files/8598511/cachegrind.out.commits.zip)
cachegrind out commits

### CODE REVIEW

**Comment by deedy5** [comment]
### line_profiler test before changes:
```
kernprof -lv performance.py
```
```
Total time: 3.58387 s
File: /charset_normalizer-master/charset_normalizer/md.py
Function: feed at line 140

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   140                                               @profile
   141                                               def feed(self, character: str) -> None:
   142   4654032    1089618.0      0.2     30.4          if (
   143   2494830     899525.0      0.4     25.1              character.isspace() is False  # includes \n \t \r \v
   144   2144442     721900.0      0.3     20.1              and character.isprintable() is False
   145     14760       4099.0      0.3      0.1              and character != "\x1A"  # Why? Its the ASCII substitute character.
   146                                                   ):
   147     14585       4963.0      0.3      0.1              self._unprintable_count += 1
   148   2494830     863770.0      0.3     24.1          self._character_count += 1

```
---
### line_profiler test after changes:
```
kernprof -lv performance.py
```
```
Total time: 1.62655 s
File: /charset_normalizer-enhanced/charset_normalizer/md.py
Function: feed at line 141

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   141                                               @profile
   142                                               def feed(self, character: str) -> None:
   143   2494830     821098.0      0.3     50.5          if is_unprintable(character):
   144     14585       4548.0      0.3      0.3              self._unprintable_count += 1
   145   2494830     800900.0      0.3     49.2          self._character_count += 1

Total time: 0.008891 s
File: /charset_normalizer-enhanced/charset_normalizer/utils.py
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool:
   204     14995       3686.0      0.2     41.5      if (
   205      7354       2635.0      0.4     29.6          character.isspace() is False  # includes \n \t \r \v
   206      7338       2419.0      0.3     27.2          and character.isprintable() is False
   207       303         84.0      0.3      0.9          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208                                                   ):
   209       302         67.0      0.2      0.8          return True
```

**Comment by deedy5** [comment]
### Before commits and mypyc compilation
```
--> Charset-Normalizer Conclusions
   --> Avg: 0.0239454577s
   --> 99th: 0.242043105s
   --> 95th: 0.130321793s
   --> 50th: 0.009998223s
```
### After commits and mypyc compilation
```
--> Charset-Normalizer Conclusions
   --> Avg: 0.010750019589361701s
   --> 99th: 0.112227622s
   --> 95th: 0.054715067s
   --> 50th: 0.00445089s
```

**Comment by deedy5** [comment]
### Tests `/bin/performance.py` without dir char-dataset/None

---
### 1) size-increase = 1
```
python3 performance.py -s 1
```
Before
```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.03203997867803838s
   --> 99th: 0.34011s
   --> 95th: 0.16578s
   --> 50th: 0.01472s
```
After commits and mypyc compilation
```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.01026181236673774s
   --> 99th: 0.10958s
   --> 95th: 0.05292s
   --> 50th: 0.00413s
```
---
### 2) size-increase = 10
```
python3 performance.py -s 10
```
Before

```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.02120435064935065s
   --> 99th: 0.42033s
   --> 95th: 0.08811s
   --> 50th: 0.00393s
```
After commits and mypyc compilation
```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.0105862987012987s
   --> 99th: 0.13491s
   --> 95th: 0.03586s
   --> 50th: 0.00137s
```
---
### 3) size-increase = 100
```
python3 performance.py -s 100
```
Before

```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.05665621212121212s
   --> 99th: 1.70197s
   --> 95th: 0.14022s
   --> 50th: 0.00617s
```
After commits and mypyc compilation
```python3
--> Charset-Normalizer Conclusions
   --> Avg: 0.04496573593073593s
   --> 99th: 1.64798s
   --> 95th: 0.1104s
   --> 50th: 0.0037s
```

**Comment by deedy5** [comment]
I also made a heat map with vprof.

***heatmap_test.py***
```python3
from glob import glob
from os.path import isdir
from charset_normalizer import detect

def get_data():
    if not isdir("./char-dataset"):
        print("This script require https://github.com/Ousret/char-dataset to be cloned on package root directory")
        exit(1)
    data = []
    for tbt_path in sorted(glob("./char-dataset/**/*.*")):
        with open(tbt_path, "rb") as fp:
            content = fp.read()
            data.append(content)
    return data

def performance_compare(size_coeff):
    data = get_data()
    len_data = len(data)
    for i,d in enumerate(data, start=1):
        d = d[:750]
        detect(d)
        print(f"{i}/{len_data}")

if __name__ == "__main__":
    performance_compare(1)
```
***vprof test***
```
vprof -c h heatmap_test.py
```
***result***
[vprof_commited.zip](https://github.com/Ousret/charset_normalizer/files/8599370/vprof_commited.zip)

There are no more bottlenecks. Speeding up is now a matter of rewriting code in cython.

File: charset_normalizer/md.py [hunk 1]
```diff
@@ -445,7 +442,7 @@ def ratio(self) -> float:
 
         return self._successive_upper_lower_count_final / self._character_count
 
-
+@lru_cache(maxsize=1024)
```
Comment (Ousret) [inline charset_normalizer/md.py:445]: I don't think that is necessary. In theory that should be even slower.

File: charset_normalizer/utils.py [hunk 2]
```diff
@@ -73,6 +73,7 @@ def is_latin(character: str) -> bool:
     return "LATIN" in description
 
 
+@lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
```
Comment (Ousret) [inline charset_normalizer/utils.py:76]: Let's not increase RAM usage when it's not really necessary.

File: setup.py [hunk 3]
```diff
@@ -51,6 +53,18 @@ def get_version():
...
+        'charset_normalizer/__init__.py',
+        'charset_normalizer/api.py',
+        'charset_normalizer/constant.py',
+        'charset_normalizer/cd.py',
+        'charset_normalizer/md.py',
+        'charset_normalizer/models.py',
+        'charset_normalizer/utils.py',
+        'charset_normalizer/assets/__init__.py',
+        'charset_normalizer/cli/normalizer.py'
+
```
Comment (Ousret) [inline setup.py:66]: extraneous line

**Review by Ousret — CHANGES_REQUESTED** [review]
I have put some through in the originally opened issue. Plus some minor remark here.

File: charset_normalizer/utils.py (hunk 2 above)
Comment (deedy5) [reply charset_normalizer/utils.py:76]: without lru_cache ([cachegrind.out.original.zip](https://github.com/Ousret/charset_normalizer/files/8598533/cachegrind.out.original.zip))

without lru_cache
---
with lru_cache ([cachegrind.out.commits.zip](https://github.com/Ousret/charset_normalizer/files/8598511/cachegrind.out.commits.zip))

with lru_cache
---

File: charset_normalizer/md.py (hunk 1 above)
Comment (deedy5) [reply charset_normalizer/md.py:445]: This function is the most callable, the effect of adding lru_cache is huge.

without lru_cache ([cachegrind.out.original.zip](https://github.com/Ousret/charset_normalizer/files/8598533/cachegrind.out.original.zip))
without lru_cache

---
with lru_cache ([cachegrind.out.commits.zip](https://github.com/Ousret/charset_normalizer/files/8598511/cachegrind.out.commits.zip))
with lru_cache

Comment (Ousret) [reply charset_normalizer/md.py:446]: And for the RAM usage increase?

Comment (deedy5) [reply charset_normalizer/md.py:446]: ### Memory test (memory-profiler)
ptest.py
```python3
from glob import glob
from os.path import isdir
from charset_normalizer import detect

def performance_compare(size_coeff):
    if not isdir("./char-dataset"):
        print("This script require https://github.com/Ousret/char-dataset to be cloned on package root directory")
        exit(1)
    for tbt_path in sorted(glob("./char-dataset/**/*.*")):
        with open(tbt_path, "rb") as fp:
            content = fp.read() * size_coeff            
        detect(content)

if __name__ == "__main__":
    performance_compare(1)
```
command
```
mprof run ptest.py
mprof plot
```
1. without lru_cache
1 without lru_cache

2. lru_cache(is_suspiciously_successive_range)
2  lru_cache(is_suspiciously_successive_range)

3. lru_cache(is_ascii)
3  lru_cache(is_ascii)

4. lru_cache(is_ascii + is_suspiciously_successive_range)
4  lru_cache(is_ascii + is_suspiciously_successive_range)

Comment (deedy5) [reply charset_normalizer/md.py:446]: ### size_coeff=100
1. without lru_cache
100 1  without lru_cache

2. lru_cache(is_ascii + is_suspiciously_successive_range)
100 4  lru_cache(is_ascii + is_suspiciously_successive_range)

Comment (Ousret) [reply charset_normalizer/md.py:446]: So the RAM usage does not change much. OK.

**Review by deedy5 — COMMENTED** [review]
Ok

**Comment by deedy5** [comment]
I did my tests and filled in the table, the performance increased by 20% with size_coeff = 1
[ptest.zip](https://github.com/Ousret/charset_normalizer/files/8602021/ptest.zip)
[difference.xlsx](https://github.com/Ousret/charset_normalizer/files/8602022/difference.xlsx)

**Comment by deedy5** [comment]
***Performance comparison table***

screen
[performance_comparison.xlsx](https://github.com/Ousret/charset_normalizer/files/8602905/performance_comparison.xlsx)

**Review by Ousret — APPROVED** [review]
LGTM

**Comment by deedy5** [comment]
+1 small correction

**Comment by deedy5** [comment]
I'm done.

**Comment by deedy5** [comment]
I'm done.

File: charset_normalizer/utils.py [hunk 4]
```diff
@@ -197,6 +198,17 @@ def is_unicode_range_secondary(range_name: str) -> bool:
     return any(keyword in range_name for keyword in UNICODE_SECONDARY_RANGE_KEYWORD)
 
 
+@lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
+def is_unprintable(character: str) -> bool:
+    if (
```
Comment (Ousret) [inline charset_normalizer/utils.py:203]: Why not simply inline directly with a return statement?

Comment (deedy5) [reply charset_normalizer/utils.py:203]: The function is cached, and there is almost no difference in performance. I leave the choice to you.
```python3
Total time: 0.010625 s                                                                                
File: /home/user/Downloads/charset_normalizer-master/charset_normalizer/utils.py                      
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool:
   204     14995       3538.0      0.2     33.3      if (
   205      7354       3100.0      0.4     29.2          character.isspace() is False  # includes \n \t \r \v
   206      7338       2261.0      0.3     21.3          and character.isprintable() is False
   207       303         78.0      0.3      0.7          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208                                               ):
   209       302         67.0      0.2      0.6          return True
   210      7052       1581.0      0.2     14.9      return False
```
```python3
Total time: 0.008757 s
File: /home/user/Downloads/charset_normalizer-master/charset_normalizer/utils.py
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool | None:
   204      7354       1456.0      0.2     16.6      return (
   205     14692       4777.0      0.3     54.6          character.isspace() is False  # includes \n \t \r \v
   206      7338       2442.0      0.3     27.9          and character.isprintable() is False
   207       303         82.0      0.3      0.9          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208                                               )
```
```python3
Total time: 0.011836 s
File: /home/user/Downloads/charset_normalizer-master/charset_normalizer/utils.py
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool:
   204     22047       5027.0      0.2     42.5      return True if (
   205      7354       3031.0      0.4     25.6          character.isspace() is False  # includes \n \t \r \v
   206      7338       2245.0      0.3     19.0          and character.isprintable() is False
   207       303         74.0      0.2      0.6          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208      7052       1459.0      0.2     12.3      ) else False
```

Comment (deedy5) [reply charset_normalizer/utils.py:203]:
```python3
Total time: 0.010625 s                                                                                
File: /charset_normalizer-master/charset_normalizer/utils.py                      
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool:
   204     14995       3538.0      0.2     33.3      if (
   205      7354       3100.0      0.4     29.2          character.isspace() is False  # includes \n \t \r \v
   206      7338       2261.0      0.3     21.3          and character.isprintable() is False
   207       303         78.0      0.3      0.7          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208                                               ):
   209       302         67.0      0.2      0.6          return True
   210      7052       1581.0      0.2     14.9      return False

```
```python3
Total time: 0.008766 s
File: /charset_normalizer-master/charset_normalizer/utils.py
Function: is_unprintable at line 201

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   201                                           @lru_cache(maxsize=UTF8_MAXIMAL_ALLOCATION)
   202                                           @profile
   203                                           def is_unprintable(character: str) -> bool:
   204      7354       1462.0      0.2     16.7      return (
   205     14692       4697.0      0.3     53.6          character.isspace() is False  # includes \n \t \r \v
   206      7338       2530.0      0.3     28.9          and character.isprintable() is False
   207       303         77.0      0.3      0.9          and character != "\x1A"  # Why? Its the ASCII substitute character.
   208                                               )
```

Comment (Ousret) [reply charset_normalizer/utils.py:203]: In favor of the 2nd.

Comment (deedy5) [reply charset_normalizer/utils.py]: Done

**Comment by Ousret** [comment]
Everything looks good. Let's proceed/move on.

A bit of caution with the expressed delays by GHA. We never know what VM is behind or how scaled.
Consider using factor between chardet and charset-normalizer.

**Comment by deedy5** [comment]
> Everything looks good. Let's proceed/move on.
> 
> A bit of caution with the expressed delays by GHA. We never know what VM is behind or how scaled. Consider using factor between chardet and charset-normalizer.
### Performance comparison
***181*** : https://github.com/Ousret/charset_normalizer/pull/181 / [performance test](https://github.com/Ousret/charset_normalizer/runs/6211035114?check_suite_focus=true)
***183*** : https://github.com/Ousret/charset_normalizer/pull/183 / [performance test](https://github.com/Ousret/charset_normalizer/runs/6277552469?check_suite_focus=true)
***Summary***
performance_comparison
[performance_comparison.xlsx](https://github.com/Ousret/charset_normalizer/files/8613938/performance_comparison.xlsx)

**Pull request merged**